language/bytecode-verifier/src/dependencies.rs

```diff
@@ -338,3 +338,48 @@ impl<'a> DependencyChecker<'a> {
         }
     }
 }
+
+pub struct CyclicModuleDependencyChecker {}
+
+impl CyclicModuleDependencyChecker {
+    pub fn verify_module<F: Fn(&ModuleId) -> Vec<ModuleId>>(
+        module: &CompiledModule,
+        immediate_module_dependencies: F,
+    ) -> VMResult<()> {
+        Self::verify_module_impl(module, immediate_module_dependencies)
+            .map_err(|e| e.finish(Location::Module(module.self_id())))
+    }
+
+    fn verify_module_impl<F: Fn(&ModuleId) -> Vec<ModuleId>>(
+        module: &CompiledModule,
+        immediate_module_dependencies: F,
+    ) -> PartialVMResult<()> {
+        // A standard DFS algorithm that follows the dependency chain from the `cursor_module_id`
+        // to see whether the `target_module_id` is in its dependencies.
+        fn is_reachable_via_dependencies<F: Fn(&ModuleId) -> Vec<ModuleId>>(
```

Comment:
I don't see the visited set optimization being used (or am I mistaken?).  You could get exponential behavior without this optimization.

Reply:
Right, I see. I'll give it a tweak.